errors.go

```diff
@@ -0,0 +1,388 @@
+func IsCompareFailed(e error) bool {
+	type cf interface {
+		IsCompareFailedError() bool
+	}
+	_, ok := Unwrap(e).(cf)
+	return ok
+}
+
+// AccessDenied returns new access denied error
+func AccessDenied(message string, args ...interface{}) error {
+	return wrap(&AccessDeniedError{
+		Message: fmt.Sprintf(message, args...),
+	}, 2)
+}
+
+// AccessDeniedError indicates denied access
+type AccessDeniedError struct {
+	Message string `json:"message"`
+}
+
+// Error is debug - friendly error message
+func (e *AccessDeniedError) Error() string {
+	if e.Message != "" {
+		return e.Message
+	}
+	return "access denied"
+}
+
+// IsAccessDeniedError indicates that this error is of AccessDenied type
+func (e *AccessDeniedError) IsAccessDeniedError() bool {
+	return true
+}
+
+// OrigError returns original error (in this case this is the error itself)
+func (e *AccessDeniedError) OrigError() error {
+	return e
+}
+
+// IsAccessDenied detects if this error is of AccessDeniedError
+func IsAccessDenied(e error) bool {
+	type ad interface {
+		IsAccessDeniedError() bool
+	}
+	_, ok := Unwrap(e).(ad)
+	return ok
+}
+
+// ConvertSystemError converts system error to appropriate teleport error
+// if it is possible, otherwise, returns original error
+func ConvertSystemError(err error) error {
+	innerError := Unwrap(err)
+
+	if os.IsNotExist(innerError) {
+		return wrap(&NotFoundError{Message: innerError.Error()}, 2)
+	}
+	if os.IsPermission(innerError) {
+		return wrap(&AccessDeniedError{Message: innerError.Error()}, 2)
+	}
+	switch realErr := innerError.(type) {
+	case *net.OpError:
+		return wrap(&ConnectionProblemError{
+			Message: fmt.Sprintf("failed to connect to server %v", realErr.Addr),
+			Err:     realErr}, 2)
+	case *os.PathError:
+		return wrap(&AccessDeniedError{
+			Message: fmt.Sprintf("failed to execute command %v error:  %v", realErr.Path, realErr.Err),
+		}, 2)
+	case x509.SystemRootsError, x509.UnknownAuthorityError:
+		return wrap(&TrustError{Err: innerError}, 2)
+	default:
+		return err
+	}
+}
+
+// ConnectionProblem returns ConnectionProblem
+func ConnectionProblem(err error, message string, args ...interface{}) error {
+	return wrap(&ConnectionProblemError{
+		Message: fmt.Sprintf(message, args...),
+		Err:     err,
+	}, 2)
+}
+
+// ConnectionProblemError indicates any network error that has occured
+type ConnectionProblemError struct {
+	Message string `json:"message"`
+	Err     error  `json:"-"`
+}
+
+// Error is debug - friendly error message
+func (c *ConnectionProblemError) Error() string {
+	return fmt.Sprintf("%v: %v", c.Message, c.Err)
+}
+
+// IsConnectionProblemError indicates that this error is of ConnectionProblem
+func (c *ConnectionProblemError) IsConnectionProblemError() bool {
+	return true
+}
+
+// OrigError returns original error (in this case this is the error itself)
+func (c *ConnectionProblemError) OrigError() error {
+	return c
+}
+
+// IsConnectionProblem detects if this error is of ConnectionProblemError
```

Comment:
`detects if` -> `returns whether` (for consistency) (from elsewhere)
`... is of ConnectionProblemError type`